### ISSUE TITLE
Numerical instability in jax.nn.log_softmax

### ISSUE THREAD

**Issue by vladfi1** [google/jax#13529]
### Description

I've stumbled upon a numerics issue in the gradients of jax.nn.log_softmax. I have a colab repro [here](https://colab.research.google.com/gist/vladfi1/c71b1f673f840b4a054b92ad870e637b/log_softmax_test.ipynb). The issue is worse for bf16 but is also noticeable in float32, and can be mitigated for both bf16 and f32 with a small change to the softmax implementation (removing the stop_gradient). I've tested this on both CPU and TPU. More experimentation [here](https://colab.research.google.com/gist/vladfi1/57e85fd0836213623965e75dd1f2708a/log_softmax_numerics_dev.ipynb) has some findings on when this issue arises, but I don't have a mathematical explanation of why it happens.

### What jax/jaxlib version are you using?

internal

### Which accelerator(s) are you using?

CPU/TPU

**Comment by levskaya** [comment]
You've just created a degenerate situation that amplifies exact cancellation errors... I'm not sure there's any easy way to fix this.

Your loss construction is pretty weird, maybe you meant to write a cross-entropy along last axis depth `n` rather than the tiny `R`?  This isn't really important though.

You're providing an extremely degenerate all-zeros logit case and seeing cancellation imprecisions in bf16 (and even f32), and then amplifying this deviation through a bunch of summed-up constant terms elsewhere in the gradient:

```python
R = 3

# focus on just the max-shift + log-softmax VJP terms:

# degenerate case exp_logits is just ones
exp_logits = jnp.ones((R,), jnp.bfloat16)
# log softmax kronecker term for first entry
knonecker = jnp.zeros((R,), jnp.bfloat16).at[0].set(1.0)
# derivative of UNSHIFTED degenerate log_softmax
z = -(knonecker - exp_logits / jnp.sum(exp_logits))
# the sum of the above vector is supposed to be zero exactly... but we're summing imprecise fractions to achieve that.
# for R=3,7, etc. the floating point sum does not cancel exactly!
# for R=2^N it will cancel exactly, and your example has no discrepancy between sg / no-sg
print(jnp.sum(z))

# further derivative passing through x - max(x) function:
# truncated not-cancelled term from max gradient tweaks the numbers:
w = z - jnp.sum(z) / R

# difference between SG and no SG version is just:
print(z - w)

# the rest of your gradient setup is just amplifying this cancellation error
# via the np.arange terms - (note sometimes for larger these exceed bf16 summation 
# range as well and cause both SG and no-SG calculations to deviate from zero)
```

**Comment by vladfi1** [comment]
To help me understand, what is the gradient of `jnp.max` in case of an n-way tie? Do each of the tied components receive 1/n'th of the incoming gradient?

**Comment by levskaya** [comment]
Yeah, `jnp.max` uses "smearing" over ties behavior, so:

```python
jax.grad(jnp.max)(jnp.array([1.0, 0.0, 0.0], dtype=jnp.bfloat16))  # DeviceArray([1, 0, 0], dtype=bfloat16)
jax.grad(jnp.max)(jnp.array([1.0, 1.0, 0.0], dtype=jnp.bfloat16))  # DeviceArray([0.5, 0.5, 0], dtype=bfloat16)
jax.grad(jnp.max)(jnp.array([0.0, 0.0, 0.0], dtype=jnp.bfloat16))  # DeviceArray([0.333984, 0.333984, 0.333984], dtype=bfloat16)
# higher precision:
jax.grad(jnp.max)(jnp.array([0.0, 0.0, 0.0], dtype=jnp.float32))   # DeviceArray([0.33333334, 0.33333334, 0.33333334], dtype=float32)
```

but this is one of the reasons you want to use a `stop_gradient` on a `max` as I believe the mixing of ties can lead to undesirable gradient signal mixing in general.

**Comment by vladfi1** [comment]
Actually it seems to me that the `-jnp.sum(z) / R` term from the fully-smeared max gradient in general moves `jnp.sum(w)` closer to 0, where it should be. Even when not fully-smeared, it will have this effect, as the sum of the max-grad will equal the sum of the incoming grad. However, this may not always be the optimal way to adjust the softmax gradient, and something more explicit might be better.

**Comment by vladfi1** [comment]
Would you consider always subtracting `jnp.sum(z) / R` from the gradient in the backwards pass? This has the effect of projecting down onto the subspace of zero-sum (gradient) vectors.

**Comment by levskaya** [comment]
I forgot that you can actually format math on github nowadays, so let me clarify the situation for myself if no one else:

We want to know the numerics of the "vjp" function of $\textrm{logsoftmax}(\textrm{maxshift}(x))$  Let's work outside-in as one does for a VJP:

Abbreviate softmax with $\sigma$

$$
\textrm{softmax}(\bar{x})_i = \sigma_i = \frac{\exp(x_i)}{\sum_k \exp(x_k)}
$$

then the jacobian of log-softmax is

$$
\frac{\partial}{\partial x_j} \log(\sigma_i) = \frac{\partial}{\partial x_j} \left[x_i - \log\left( \sum_k \exp(x_k) \right)\right] = \delta_{ij} - \sigma_j
$$

so the reverse-mode formula for transforming gradients is

$$
g_i \rightarrow g_i - \left(\sum_k g_k\right) \sigma_i    \tag{eq1}
$$

Now, for the "max-shift" function, let's first consider that we have $\max$ function with smear over ties gradient behavior:

$$
\frac{\partial}{\partial x_j} \max(\bar{x}) = 
\begin{cases}
  \left(\textrm{num ties}\right)^{-1} & \text{if } x_j = \max(x)\\
  0         & \text{otherwise}
\end{cases}
$$

we can write the derivative in terms of iverson bracket syntax:

$$
\frac{\partial}{\partial x_j} \max(\bar{x}) = \frac{\left[\\!\left[x_j = max(x)\right]\\!\right]}{\sum_i \left[\\!\left[x_i = max(x)\right]\\!\right]} = \mu_j
$$

then the jacobian is

$$
\frac{\partial}{\partial x_j} \left( x_i - \max(\bar{x}) \right) = \delta_{ij} - \mu_j
$$

and the reverse-mode formula is

$$
g_i \rightarrow g_i - \left( \sum_k g_k \right) \mu_i    \tag{eq2}
$$

much like the log-softmax formula.  Note that it is this $\left( \sum_k g_k \right) \mu_i$ term that is _eliminated_ if we use a "stop-gradient" on the max function.

The important part of your setup starts by only considering the first output component, this is the same as thinking about the whole VJP, starting w. a gradient vector:

$$
g_0  = \left( 1 ,\\; 0 ,\\; 0 \right)
$$

Using (eq1) we have (since $\sigma_i = \frac{1}{3}$):

$$
g_1  = \left( 1 - \frac{1}{3} ,\\; - \frac{1}{3} ,\\; - \frac{1}{3} \right)
$$

And via (eq2) we have the "stop-gradient term" (since $\mu_i = \frac{1}{3}$):

$$
\frac{1}{3} \left( \left(1 - \frac{1}{3} \right) - \frac{1}{3}  - \frac{1}{3} \right)
$$

This is nonzero in bfloat16 because of precision loss (but note that $1 - \left(\frac{1}{3} - \frac{1}{3}  - \frac{1}{3}\right)$ is zero in bfloat16).

So, yes, the "max-gradient" term that JAX omits is "useful" in this case in projecting to the zero-sum grad-subspace... 
 but I'm not sure that's generally useful?  or even desired?

If I recall, I believe that the worry is that it can also strongly mix the backprop gradients corresponding to tied-max inputs in a way that's nearly identical/redundant to the mixing they would have just undergone in the softmax fn... but practically it also just saves a bit of compute, and away from these degenerate cases it's unclear if there's any benefit.
